### PR TITLE
Magnetic field declination table bounds check

### PR DESCRIPTION
**Pull request by stephanbro**
Fixes an issue where the upper bound of the declination table wasn't actually enforced. Luckily, it's only an issue if using PX4 north of the Arctic Circle!

This PR also updates the declination values based on the 2015 magnetic field data from [NOAA](https://www.ngdc.noaa.gov/geomag/WMM/) using the script here: https://github.com/Skycatch/geomag. The data also provides a prediction about the Earth's magnetic field and could be used to provide values for today's date based on that prediction.

### CODE REVIEW

File: src/lib/geo_lookup/geo_mag_declination.c [hunk 1]
```diff
@@ -54,22 +55,41 @@
...
 static float get_lookup_table_val(unsigned lat, unsigned lon);
+static unsigned get_lookup_table_index(float val, float min, float max);
+
+unsigned get_lookup_table_index(float val, float min, float max)
+{
+	/* for the rare case of hitting the bounds exactly
+	 * the rounding logic wouldn't fit, so enforce it.
+	 */
+	/* limit to table bounds - required for maxima even when table spans full globe range */
+	if (val <= min) {
```
Comment (LorenzMeier) [inline src/lib/geo_lookup/geo_mag_declination.c]: Why do we compare to smaller / equal and then set to min? Doesn't it make more sense to compare to smaller than min and set to it?

**Comment by LorenzMeier** [comment]
@dagar What are our plans with GTest?

**Comment by LorenzMeier** [comment]
@stephanbro Nice, thanks!

File: src/systemcmds/tests/test_autodeclination.cpp [hunk 2]
```diff
@@ -23,7 +23,7 @@ class AutoDeclinationTest : public UnitTest
 
 bool AutoDeclinationTest::autodeclination_check(void)
 {
-	ut_assert("declination differs more than 1 degree", get_mag_declination(47.0, 8.0) - 0.6f < 0.5f);
+	ut_assert("declination differs more than 1 degree", get_mag_declination(47.0, 8.0) - 1.6f < 0.5f);
```
Comment (LorenzMeier) [inline src/systemcmds/tests/test_autodeclination.cpp]: @stephanbro We run this test suite mostly because it runs in SITL, in CI as unit test and on-hardware. Could you add all test cases you added to GTest here as well? GTest might be eventually factored out.

File: src/lib/geo_lookup/geo_mag_declination.c (hunk 1 above)
Comment (stephanbro) [inline src/lib/geo_lookup/geo_mag_declination.c]: I don't see why we couldn't just use a `<` instead, as you suggested. I was just matching the existing style here: https://github.com/PX4/Firmware/pull/6503/files#diff-9c5c63f21338acbe6fedb1f1f35f9cfeL95

File: src/systemcmds/tests/test_autodeclination.cpp (hunk 2 above)
Comment (stephanbro) [inline src/systemcmds/tests/test_autodeclination.cpp]: The issue is that in the gtest I'm including the `geo_mag_declination.c` file so I have access to the static methods in the file. Because that file is already built in SITL there will be multiple definitions of functions if I include the same file in this test. The only way I see to make my tests work would be to move the function declaration to the .h file, but that doesn't make sense from an encapsulation perspective. Any ideas?

Comment (LorenzMeier) [inline src/systemcmds/tests/test_autodeclination.cpp]: Could we just duplicate the boundary condition numbers you've used in the other test?

Comment (stephanbro) [inline src/systemcmds/tests/test_autodeclination.cpp]: Yeah I can do that, no problem.

**Comment by stephanbro** [comment]
@LorenzMeier I updated the testing and remove the gtest items. I also found and fixed an additional issue when using values outside of the sampling range. Let me know if you'd like me to rebase.

**Comment by dagar** [comment]
Right now our standalone unit tests (https://github.com/PX4/Firmware/tree/master/unittests) use gtest, but there's only a single param test in there.
The PX4 tests command (using our unit_test framework) runs within PX4 on all platforms and built with the same compilers and options as actual usage. It has absorbed almost all of the previous gtest unit tests. Worse test framework, better testing.

If desired we can still do the standalone test if you want to keep geo as a single isolated lib that could be used outside of PX4. It would be nice if other components were like that, but they tend to be dependant on other parts of PX4.

I tried to get gtest working within Nuttx (to replace our own unit test framework), but failed due to a seemingly endless list of c++ issues. I was considering giving it another try since we've updated Nuttx.

**Comment by LorenzMeier** [comment]
Applied, thanks!
